nexus/db-queries/src/db/queries/vpc_subnet.rs

```diff
@@ -7,407 +7,318 @@
-                    format!("IP address range '{}' conflicts with an existing subnet", ip).as_str()
-                )
-            },
-            SubnetError::External(e) => e,
-        }
-    }
-}
-
-/// Generate a subquery that selects any overlapping address ranges of the same
-/// type as the input IP subnet.
+/// Query used to insert VPC Subnets.
 ///
-/// This generates a query that, in full, looks like:
+/// This query is used to idempotently insert a VPC Subnet. The query also looks
+/// for any other subnets in the same VPC whose IP address blocks overlap. All
+/// Subnets are required to have non-overlapping IP blocks.
 ///
-/// ```sql
-/// SELECT
-///     <ip>
-/// FROM
-///     vpc_subnet
-/// WHERE
-///     vpc_id = <vpc_id> AND
-///     time_deleted IS NULL AND
-///     inet_contains_or_equals(ipv*_block, <ip>)
-/// LIMIT 1
-/// ```
+/// Note that this query is idempotent. If a record with the provided primary
+/// key already exists, that record is returned exactly from the DB, without any
+/// other modification or alteration. If callers care, they can inspect the
+/// record to make sure it's what they expected, though that's usually a fraught
+/// endeavor.
 ///
-/// The input may be either an IPv4 or IPv6 subnet, and the corresponding column
-/// is compared against. Note that the exact input IP range is returned on
-/// purpose.
-fn push_select_overlapping_ip_range<'a>(
-    mut out: AstPass<'_, 'a, Pg>,
-    vpc_id: &'a Uuid,
-    ip: &'a ipnetwork::IpNetwork,
-) -> diesel::QueryResult<()> {
-    use crate::db::schema::vpc_subnet::dsl;
-    out.push_sql("SELECT ");
-    out.push_bind_param::<sql_types::Inet, ipnetwork::IpNetwork>(ip)?;
-    out.push_sql(" FROM ");
-    VPC_SUBNET_FROM_CLAUSE.walk_ast(out.reborrow())?;
-    out.push_sql(" WHERE ");
-    out.push_identifier(dsl::vpc_id::NAME)?;
-    out.push_sql(" = ");
-    out.push_bind_param::<sql_types::Uuid, Uuid>(vpc_id)?;
-    out.push_sql(" AND ");
-    out.push_identifier(dsl::time_deleted::NAME)?;
-    out.push_sql(" IS NULL AND inet_contains_or_equals(");
-    if ip.is_ipv4() {
-        out.push_identifier(dsl::ipv4_block::NAME)?;
-    } else {
-        out.push_identifier(dsl::ipv6_block::NAME)?;
-    }
-    out.push_sql(", ");
-    out.push_bind_param::<sql_types::Inet, ipnetwork::IpNetwork>(ip)?;
-    out.push_sql(")");
-    Ok(())
-}
-
-/// Generate a subquery that returns NULL if there is an overlapping IP address
-/// range of any type.
-///
-/// This specifically generates a query that looks like:
+/// Here is the entire query:
 ///
 /// ```sql
-/// SELECT NULLIF(
-///     <ip>,
-///     push_select_overlapping_ip_range(<vpc_id>, <ip>)
+/// WITH
+/// -- This CTE generates a casting error if any live records, other than _this_
+/// -- record, have overlapping IP blocks of either family.
+/// overlap AS MATERIALIZED (
+///     SELECT
+///         -- NOTE: This cast always fails, we just use _how_ it fails to
+///         -- learn which IP block overlaps.
```

Comment:
Ah, because of the `WHERE` clause below? Clever. Might be worth an explicit call-out that we depend on that filter in this comment.

Reply:
Yeah, that's right. This cast is only called when there _is_ a record with a different PK that overlaps with one we're inserting. We would just like to know which block overlapped, to report a helpful error message. Will add more notes.

Reply:
Expanded the comment in 11197cc0a